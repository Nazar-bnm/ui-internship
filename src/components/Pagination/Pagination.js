/* eslint-disable react/no-did-update-set-state */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import './Pagination.scss';

const CN = 'pagination';

class Pagination extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isValidInput: true,
      isAnimation: false
    };

    this.setPage = this.setPage.bind(this);
    this.setCurrentPage = this.setCurrentPage.bind(this);
    this.renderNumbers = this.renderNumbers.bind(this);
    this.goToPrevPage = this.goToPrevPage.bind(this);
    this.goToNextPage = this.goToNextPage.bind(this);
    this.checkThenChangePage = this.checkThenChangePage.bind(this);
    this.onAnimationEnd = this.onAnimationEnd.bind(this);
    this.inputRef = React.createRef();
  }

  componentDidMount() {
    const { currentPage } = this.props;

    this.inputRef.current.value = currentPage;
  }

  componentDidUpdate({ currentPage: prevCurrentPage }) {
    const { currentPage, isValidInput, numberOfPages } = this.props;

    if (prevCurrentPage !== currentPage) {
      this.inputRef.current.value = currentPage;

      !isValidInput
        && this.setState({
          isValidInput: true
        });
    }

    if (numberOfPages === 1) {
      this.inputRef.current.value = 1;
    }
  }

  onAnimationEnd() {
    this.setState({
      isAnimation: false
    });
  }

  setCurrentPage(value) {
    const { setCurrentPage } = this.props;

    setCurrentPage(+value);
  }

  setPage(e) {
    this.setCurrentPage(+e.target.innerText);
  }

  goToPrevPage() {
    const { currentPage } = this.props;

    this.setCurrentPage(currentPage - 1);
  }

  goToNextPage() {
    const { currentPage } = this.props;

    this.setCurrentPage(currentPage + 1);
  }

  checkThenChangePage(e) {
    const { numberOfPages } = this.props;

    const {
      target: { value: inputValue }
    } = e;

    if (inputValue === '') {
      this.setState({
        isValidInput: true
      });
      return;
    }

    if (inputValue > numberOfPages || inputValue < 1) {
      this.setState({
        isValidInput: false,
        isAnimation: true
      });
    } else {
      this.setCurrentPage(+inputValue);

      this.setState({
        isValidInput: true
      });
    }
  }

  renderNumbers() {
    const { visibleNumbers, currentPage, numberOfPages } = this.props;
    const arrayOfPages = [...Array(numberOfPages)].map((e, i) => i + 1);
    const numbersFromCurrentPage = Math.floor(visibleNumbers / 2);
    let arrayOfVisibleNumOfPages = [...arrayOfPages];

    if (numberOfPages >= 3) {
      arrayOfVisibleNumOfPages = arrayOfPages.slice(
        currentPage - numbersFromCurrentPage - 1,
        currentPage + numbersFromCurrentPage
      );

      currentPage <= numbersFromCurrentPage
        && (arrayOfVisibleNumOfPages = arrayOfPages.slice(0, visibleNumbers));

      numberOfPages - currentPage < numbersFromCurrentPage
        && (arrayOfVisibleNumOfPages = arrayOfPages.slice(-visibleNumbers));
    }

    return arrayOfVisibleNumOfPages.map((num) => (
      <span
        key={num}
        className={cx(`${CN}-nav-number`, {
          [`${CN}-nav-number--active`]: num === currentPage
        })}
        onClick={this.setPage}
      >
        {num}
      </span>
    ));
  }

  render() {
    const { className, numberOfPages, currentPage } = this.props;
    const { isValidInput, isAnimation } = this.state;

    return (
      <div className={cx(CN, className)}>
        <div className={`${CN}-nav-input-container`}>
          <span className={`${CN}-nav-input-container__text`}>go to page</span>
          <input
            ref={this.inputRef}
            disabled={numberOfPages === 1}
            onInput={this.checkThenChangePage}
            className={cx(`${CN}-nav-input-container__input`, {
              [`${CN}-nav-input-container__input--error`]: !isValidInput,
              [`${CN}-nav-input-container__input--animation`]: isAnimation
            })}
            onAnimationEnd={this.onAnimationEnd}
            type="number"
            min="1"
            max={numberOfPages}
          />
          <span className={`${CN}-nav-input-container__text`}>
            {` / ${numberOfPages}`}
          </span>
        </div>
        {numberOfPages > 1 && (
          <div className={`${CN}-nav-numbers`}>
            <div
              onClick={this.setCurrentPage}
              className={cx(`${CN}-nav-arrow`, `${CN}-nav-arrow-left`, {
                [`${CN}-nav-arrow--disabled`]: currentPage === 1
              })}
            >
              <i className={`icon chevron left ${CN}-nav-arrow__icon`} />
            </div>
            {this.renderNumbers()}
            <div
              onClick={this.goToNextPage}
              className={cx(`${CN}-nav-arrow`, `${CN}-nav-arrow-right`, {
                [`${CN}-nav-arrow--disabled`]: currentPage === numberOfPages
              })}
            >
              <i className={`icon chevron right ${CN}-nav-arrow__icon`} />
            </div>
          </div>
        )}
      </div>
    );
  }
}

Pagination.propTypes = {
  className: PropTypes.string,
  numberOfPages: PropTypes.number.isRequired,
  currentPage: PropTypes.number,
  setCurrentPage: PropTypes.func.isRequired
};

Pagination.defaultProps = {
  className: '',
  currentPage: 1
};

export default Pagination;
