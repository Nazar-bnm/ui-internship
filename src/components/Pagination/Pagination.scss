@import '../../scss/global_styles';

.pagination {
  height: max-content;
  display: flex;
  color: $dove-gray;


  &-nav {
    &-number {
      font-family: font(primary-light);
      font-size: 1rem;
      padding: 3px 6px;
      background-color: transparent;
      transition: 0.2s ease background-color;

      &--active {
        border: 1px solid $silver-chalice;
      }
    }

    &-arrow {
      padding: 3px;
      color: $silver-chalice;

      &:nth-child(1) {
        margin-right: 3px;
      }

      &:last-child {
        margin-left: 3px;
      }

      &__icon {
        margin: 0 !important;
      }

      &--disabled {
        visibility: hidden;
      }
    }

    &-input-container {
      &__text {
        font-size: 0.7rem;
        font-family: font(primary-bold);
        font-weight: bold;
        text-transform: uppercase;
        color: $dove-gray;

        &:nth-child(1) {
          margin-right: 10px;
        }
  
        &:last-child {
          margin-left: 6px;
        }
      }

      &__input {
        padding: 5px;
        border: 1px solid $silver-chalice;
        font-family: font(primary-light);
        font-size: 0.8rem;
        color: $dove-gray;
        &--error {
          outline: 0;
          border-color: red;
        }

        &--animation {
          animation-name: bounce;
          animation-duration: 0.5s;
          animation-delay: 0.25s;
        }
      }
    }
  }

  .arr-cover {
    width: 1px;
    font-size: 20px;
    text-align: center;
  }

  .arrow {
    padding: 20px;
  }

  @keyframes bounce {
    0% {
      transform: translateX(0px);
      timing-function: ease-in;
    }
    37% {
      transform: translateX(5px);
      timing-function: ease-out;
    }
    55% {
      transform: translateX(-5px);
      timing-function: ease-in;
    }
    73% {
      transform: translateX(4px);
      timing-function: ease-out;
    }
    82% {
      transform: translateX(-4px);
      timing-function: ease-in;
    }
    91% {
      transform: translateX(2px);
      timing-function: ease-out;
    }
    96% {
      transform: translateX(-2px);
      timing-function: ease-in;
    }
    100% {
      transform: translateX(0px);
      timing-function: ease-in;
    }
  }
}
