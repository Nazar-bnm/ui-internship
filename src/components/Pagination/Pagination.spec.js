import React from "react";
import { mount } from "enzyme";

import Pagination from "./Pagination";

describe("<Pagination />", () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      currentPage: "10",
      numberOfPages: 10,
      visibleNumbers: 3,
      setCurrentPage: () => {}
    };

    wrapper = mount(<Pagination {...props} />);
  });

  it("should match snapshot <Pagination />", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should set inputValue to currentPage prop or 1", () => {
    const inputInstance = wrapper.find({ type: "number" }).instance();

    expect(inputInstance.value).toBe(wrapper.props().currentPage);
    wrapper.setProps({ currentPage: "12" });
    expect(inputInstance.value).toBe(wrapper.props().currentPage);
    wrapper.setProps({ numberOfPages: 1 });
    expect(inputInstance.value).toBe('1');
  });

  it("should set state isAnimation to false when animation ends", () => {
    wrapper.find({ type: "number" }).simulate("animationEnd");
    expect(wrapper.state().isAnimation).toBeFalsy();
  });

  it("should call setCurrent page on some dom events", () => {
    const mockEvent = {
      target: {
        value: 5
      }
    };
    const mockFunc = jest.fn();

    wrapper.setProps({setCurrentPage: mockFunc});

    wrapper.find({ type: "number" }).simulate("input", mockEvent);
    expect(mockFunc).toBeCalledTimes(1);
    
    wrapper
      .find(".pagination-nav-arrow")
      .first()
      .simulate("click");
    expect(mockFunc).toBeCalledTimes(2);

    wrapper
      .find(".pagination-nav-arrow")
      .last()
      .simulate("click");
    expect(mockFunc).toBeCalledTimes(3);

    wrapper
      .find(".pagination-nav-number")
      .first()
      .simulate("click");
    expect(mockFunc).toBeCalledTimes(4);
  });

  it("should set isValid input true or false depending on input value", () => {
    let mockEvent = {
      target: {
        value: ""
      }
    };

    expect(wrapper.state().isValidInput).toBeTruthy();

    wrapper.find({ type: "number" }).simulate("input", mockEvent);
    expect(wrapper.state().isValidInput).toBeTruthy();

    mockEvent.target.value = '15';
    wrapper.find({ type: "number" }).simulate("input", mockEvent);
    expect(wrapper.state().isValidInput).toBeFalsy();
  });

  it("should not be any navigation numbers when the number of pages is equal to 1", () => {
    wrapper.setProps({numberOfPages: 1})

    expect(wrapper.exists(".pagination-nav-numbers")).toBeFalsy();
  });
});
