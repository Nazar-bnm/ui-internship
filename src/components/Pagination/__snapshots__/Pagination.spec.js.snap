// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should match snapshot <Pagination /> 1`] = `
<Pagination
  className=""
  currentPage="10"
  numberOfPages={10}
  setCurrentPage={[Function]}
  visibleNumbers={3}
>
  <div
    className="pagination"
  >
    <div
      className="pagination-nav-input-container"
    >
      <span
        className="pagination-nav-input-container__text"
      >
        go to page
      </span>
      <input
        className="pagination-nav-input-container__input"
        disabled={false}
        max={10}
        min="1"
        onAnimationEnd={[Function]}
        onInput={[Function]}
        type="number"
      />
      <span
        className="pagination-nav-input-container__text"
      >
         / 10
      </span>
    </div>
    <div
      className="pagination-nav-numbers"
    >
      <div
        className="pagination-nav-arrow pagination-nav-arrow-left"
        onClick={[Function]}
      >
        <i
          className="icon chevron left pagination-nav-arrow__icon"
        />
      </div>
      <span
        className="pagination-nav-number"
        key="8"
        onClick={[Function]}
      >
        8
      </span>
      <span
        className="pagination-nav-number"
        key="9"
        onClick={[Function]}
      >
        9
      </span>
      <span
        className="pagination-nav-number"
        key="10"
        onClick={[Function]}
      >
        10
      </span>
      <div
        className="pagination-nav-arrow pagination-nav-arrow-right"
        onClick={[Function]}
      >
        <i
          className="icon chevron right pagination-nav-arrow__icon"
        />
      </div>
    </div>
  </div>
</Pagination>
`;
